### PR TITLE
feat(cae/component): add new resource component

### PR DESCRIPTION
**Pull request by wuzhuanhong**
**What this PR does / why we need it**:

Add a new resource used to create the CAE component.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. support a new resource used to create the component.
2. support related document and acceptance test.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
 make testacc TEST=./huaweicloud/services/acceptance/cae TESTARGS='-run TestAccComponent_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cae -v -run TestAccComponent_basic -timeout 360m -parallel 4
=== RUN   TestAccComponent_basic
=== PAUSE TestAccComponent_basic
=== CONT  TestAccComponent_basic
--- PASS: TestAccComponent_basic (35.22s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cae       35.261s

```

### CODE REVIEW

File: huaweicloud/config/endpoints.go [hunk 1]
```diff
@@ -277,6 +277,11 @@ var allServiceCatalog = map[string]ServiceCatalog{
 		Version: "v1",
 		Product: "SFSTurbo",
 	},
+	"cae": {
+		Name:    "cae",
+		Version: "v1",
+		Product: "CAE",
+	},
```
Comment (Lance52259) [inline huaweicloud/config/endpoints.go]:
```suggestion
	"cae": {
		Name:    "cae",
		Product: "CAE",
	},
```
Version is useless for auto-generator, and the location should be above `mrs`

**Comment by Lance52259** [comment]
invalid file name, want 'resource_huaweicloud_cae_component.go'

File: huaweicloud/services/acceptance/acceptance.go [hunk 2]
```diff
@@ -1465,3 +1475,11 @@ func TestAccPreCheckHSSHostProtectionQuotaId(t *testing.T) {
 		t.Skip("HW_HSS_HOST_PROTECTION_QUOTA_ID must be set for the acceptance test")
 	}
 }
+
+// lintignore:AT003
+func TestAccPreCheckCAEComponent(t *testing.T) {
+	if HW_CAE_ENVIRONMENT_ID == "" || HW_CAE_APPLICATION_ID == "" || HW_CAE_CODE_URL == "" || HW_CAE_CODE_AUTH_NAME == "" ||
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/acceptance.go]:
```
// Before the CAE environment resource is released, temporarily use this environment variable for acceptance tests.
// lintignore:AT003
func TestAccPreCheckCaeEnvironment(t *testing.T) {
	if HW_CAE_ENVIRONMENT_ID == "" {
		t.Skip("HW_CAE_ENVIRONMENT_ID must be set for the CAE acceptance test")
	}
}

// Before the CAE application resource is released, temporarily use this environment variable for acceptance tests.
// lintignore:AT003
func TestAccPreCheckCaeApplication(t *testing.T) {
	if HW_CAE_APPLICATION_ID == "" {
		t.Skip("HW_CAE_APPLICATION_ID must be set for the CAE acceptance test")
	}
}
```
Please check these two environment variables separately

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+// GetCompontentById is a method to query component details from a specified application ID using given parameters.
+func GetCompontentById(cfg *config.Config, region, applicationId, id string) (interface{}, error) {
+	var (
+		httpUrl     = "v1/{project_id}/cae/applications/{application_id}/components/{component_id}"
+		product     = "cae"
+		client, err = cfg.NewServiceClient(product, region)
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error creating CAE Client: %s", err)
+	}
```
Comment (Lance52259) [inline huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]:
```suggestion
	var (
		httpUrl     = "v1/{project_id}/cae/applications/{application_id}/components/{component_id}"
		product     = "cae"
	)
	client, err := cfg.NewServiceClient(product, region)
	if err != nil {
		return nil, fmt.Errorf("error creating CAE Client: %s", err)
	}
```

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+		}
+
+		errCode, err := jmespath.Search("error_code", err)
+		if err != nil {
+			return fmt.Errorf("error parse errorCode from response body: %s", err.Error())
+		}
+
+		if errCode == "CAE.01500404" || errCode == "CAE.01500208" {
+			return golangsdk.ErrDefault404{}
+		}
```
Comment (Lance52259) [inline huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]:
```suggestion
		// Error code CAE.01500404 indicates that the environment does not exist.
		// Error code CAE.01500208 indicates that the application or component does not exist.
		if errCode == "CAE.01500404" || errCode == "CAE.01500208" {
			return golangsdk.ErrDefault404(errCode)
		}
```

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+	importedId := d.Id()
+	parts := strings.Split(importedId, "/")
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("invalid format specified for import ID, want '<environment_id>/<application_id>/<id>', but got '%s'",
+			importedId)
+	}
+
+	d.SetId(parts[2])
+	mErr := multierror.Append(nil,
+		d.Set("application_id", parts[1]),
```
Comment (Lance52259) [inline huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]: did we forgot the environment ID setting?

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+	var err interface{}
+	if errCode, ok := respErr.(golangsdk.ErrDefault400); ok {
+		pErr := json.Unmarshal(errCode.Body, &err)
+		if pErr != nil {
+			return pErr
+		}
+
+		errCode, err := jmespath.Search("error_code", err)
+		if err != nil {
+			return fmt.Errorf("error parse errorCode from response body: %s", err.Error())
```
Comment (Lance52259) [inline huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]:
```suggestion
			return fmt.Errorf("error parse error code from the API response body: %s", err.Error())
```

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]:
```
// ParseQueryError400 is a method used to parse whether a 404 error message means the resources not found.
// For the CAE service, there are some known 404 error codes:
// + CAE.01500208: application or component does not found.
// + CAE.01500404: environment does not found.
func ParseQueryError400(err error, specErrors []string) error {
	var err400 golangsdk.ErrDefault400
	if errors.As(err, &err400) {
		var apiError interface{}
		if jsonErr := json.Unmarshal(err400.Body, &apiError); jsonErr != nil {
			return err
		}

		errCode, err := jmespath.Search("error_code", apiError)
		if errCode != nil {
			return err
		}

		for _, v := range specErrors {
			if errCode == v {
				return golangsdk.ErrDefault404{}
			}
		}
	}
	return err
}
```
It is recommend to define this function in the `common.go` for CAE service package.

File: huaweicloud/config/endpoints.go (hunk 1 above)
Comment (wuzhuanhong) [reply huaweicloud/config/endpoints.go]: fixed

File: huaweicloud/services/acceptance/acceptance.go (hunk 2 above)
Comment (wuzhuanhong) [reply huaweicloud/services/acceptance/acceptance.go]: fixed

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go (hunk 3 above)
Comment (wuzhuanhong) [reply huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]: fixed

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go (hunk 4 above)
Comment (wuzhuanhong) [reply huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]: fixed

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go (hunk 5 above)
Comment (wuzhuanhong) [reply huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]: fixed

File: huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go (hunk 6 above)
Comment (wuzhuanhong) [reply huaweicloud/services/cae/resource_huaweicloud_cae_compontent.go]: fixed

**Comment by wuzhuanhong** [comment]
make testacc TEST=./huaweicloud/services/acceptance/cae TESTARGS='-run TestAccComponent_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cae -v -run TestAccComponent_basic -timeout 360m -parallel 4
=== RUN   TestAccComponent_basic
=== PAUSE TestAccComponent_basic
=== CONT  TestAccComponent_basic
--- PASS: TestAccComponent_basic (40.71s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cae       40.757s

File: docs/resources/cae_component.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+```bash
+$ terraform import huaweicloud_cae_component.test <environment_id>/<application_id>/<id>
+```
+
+Note that the imported state may not be identical to your resource definition, due to some attributes missing from the
+API response, security or some other reason.
+The missing attributes include: `metadata.0.annotations`, `spec.0.build.0.parameters`.
+It is generally recommended running `terraform plan` after importing the resource.
+You can then decide if changes should be applied to the instance, or the resource definition should be updated to
+align with the instance. Also you can ignore changes as below.
```
Comment (Lance52259) [inline docs/resources/cae_component.md]:
```suggestion
You can then decide if changes should be applied to the resource, or the resource definition should be updated to
align with the resource. Also you can ignore changes as below.
```

File: huaweicloud/services/cae/resource_huaweicloud_cae_component.go [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+	updateComponentOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		MoreHeaders: map[string]string{
+			"X-Environment-Id": d.Get("environment_id").(string),
+		},
+		JSONBody: utils.RemoveNil(buildCreateOrUpdateComponentBodyParams(d)),
+	}
+	_, err = client.Request("PUT", updatePath, &updateComponentOpt)
+	if err != nil {
+		return diag.Errorf("error updating CAE component (%s): %s", d.Id(), err)
```
Comment (Lance52259) [inline huaweicloud/services/cae/resource_huaweicloud_cae_component.go]:
```suggestion
	updatePath = strings.ReplaceAll(updatePath, "{component_id}", componentId)

	updateComponentOpt := golangsdk.RequestOpts{
		KeepResponseBody: true,
		MoreHeaders: map[string]string{
			"X-Environment-Id": d.Get("environment_id").(string),
		},
		JSONBody: utils.RemoveNil(buildCreateOrUpdateComponentBodyParams(d)),
	}
	_, err = client.Request("PUT", updatePath, &updateComponentOpt)
	if err != nil {
		return diag.Errorf("error updating CAE component (%s): %s", componentId, err)
```
suggest to define a variable and names 'componentId'

**Comment by wuzhuanhong** [comment]
make testacc TEST=./huaweicloud/services/acceptance/cae TESTARGS='-run TestAccComponent_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cae -v -run TestAccComponent_basic -timeout 360m -parallel 4
=== RUN   TestAccComponent_basic
=== PAUSE TestAccComponent_basic
=== CONT  TestAccComponent_basic
--- PASS: TestAccComponent_basic (35.64s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cae       35.690s

**Pull request merged**